File: engineering/apps/games/sudoku/sudoku.cpp
#include "sudoku.h"

/* 宏别名，确保代码兼容 */
#define BOARD_SIZE SUDOKU_BOARD_SIZE
#define BOX_SIZE SUDOKU_BOX_SIZE

/* ========== 难度配置 ========== */
const DifficultyConfig DIFFICULTIES[] = {
    { "简单", 30 },
    { "中等", 40 },
    { "困难", 50 },
};

/* ========== 平台适配（复用 snake 项目模式） ========== */

#ifdef _WIN32
static HANDLE hConsole;
static HANDLE hOutput;
static DWORD  oldMode;
static DWORD  oldOutputMode;
#endif

void set_utf8(void) {
#ifdef _WIN32
    SetConsoleOutputCP(65001);
    SetConsoleCP(65001);
#endif
}

void sleep_ms(int ms) {
#ifdef _WIN32
    Sleep(ms);
#else
    usleep((unsigned int)ms * 1000);
#endif
}

int my_getch(void) {
#ifdef _WIN32
    return _getch();
#else
    struct termios oldt, newt;
    tcgetattr(0, &oldt);
    newt = oldt;
    newt.c_lflag &= ~(ICANON | ECHO);
    tcsetattr(0, TCSANOW, &newt);
    int c = getchar();
    tcsetattr(0, TCSANOW, &oldt);
    return c;
#endif
}

int my_kbhit(void) {
#ifdef _WIN32
    return _kbhit();
#else
    struct timeval tv = {0, 0};
    fd_set readset;
    FD_ZERO(&readset);
    FD_SET(0, &readset);
    return select(1, &readset, NULL, NULL, &tv) > 0;
#endif
}

void console_raw_mode(void) {
#ifdef _WIN32
    hConsole = GetStdHandle(STD_INPUT_HANDLE);
    GetConsoleMode(hConsole, &oldMode);
    SetConsoleMode(hConsole, oldMode & ~(ENABLE_LINE_INPUT | ENABLE_ECHO_INPUT));

    hOutput = GetStdHandle(STD_OUTPUT_HANDLE);
    GetConsoleMode(hOutput, &oldOutputMode);
    DWORD dwMode = oldOutputMode | ENABLE_VIRTUAL_TERMINAL_PROCESSING;
    SetConsoleMode(hOutput, dwMode);
#endif
}

void console_restore(void) {
#ifdef _WIN32
    SetConsoleMode(hConsole, oldMode);
    SetConsoleMode(hOutput, oldOutputMode);
#endif
}

/* ========== 核心算法 ========== */

/* 检查在 (row,col) 填入 num 是否合法 */
bool is_valid_placement(const Cell board[BOARD_SIZE][BOARD_SIZE],
                        int row, int col, int num) {
    /* 行检查 */
    for (int c = 0; c < BOARD_SIZE; c++) {
        if (c != col && board[row][c].value == num) return false;
    }
    /* 列检查 */
    for (int r = 0; r < BOARD_SIZE; r++) {
        if (r != row && board[r][col].value == num) return false;
    }
    /* 3×3 宫检查 */
    int box_r = row / BOX_SIZE * BOX_SIZE;
    int box_c = col / BOX_SIZE * BOX_SIZE;
    for (int r = box_r; r < box_r + BOX_SIZE; r++) {
        for (int c = box_c; c < box_c + BOX_SIZE; c++) {
            if ((r != row || c != col) && board[r][c].value == num)
                return false;
        }
    }
    return true;
}

/* 回溯求解器：找到第一个空格，依次尝试 1-9 */
bool solve_sudoku(Cell board[BOARD_SIZE][BOARD_SIZE]) {
    for (int r = 0; r < BOARD_SIZE; r++) {
        for (int c = 0; c < BOARD_SIZE; c++) {
            if (board[r][c].value == 0) {
                for (int num = 1; num <= 9; num++) {
                    if (is_valid_placement(board, r, c, num)) {
                        board[r][c].value = num;
                        if (solve_sudoku(board)) return true;
                        board[r][c].value = 0;
                    }
                }
                return false; /* 1-9 都不行，回溯 */
            }
        }
    }
    return true; /* 没有空格了，求解成功 */
}

/* 计数求解器：统计解的数量，达到 limit 提前返回 */
int count_solutions(Cell board[BOARD_SIZE][BOARD_SIZE], int limit) {
    for (int r = 0; r < BOARD_SIZE; r++) {
        for (int c = 0; c < BOARD_SIZE; c++) {
            if (board[r][c].value == 0) {
                int count = 0;
                for (int num = 1; num <= 9; num++) {
                    if (is_valid_placement(board, r, c, num)) {
                        board[r][c].value = num;
                        count += count_solutions(board, limit);
                        board[r][c].value = 0;
                        if (count >= limit) return count;
                    }
                }
                return count;
            }
        }
    }
    return 1; /* 没有空格了，找到 1 个解 */
}

/* Fisher-Yates 洗牌 */
static void shuffle(int arr[], int n) {
    for (int i = n - 1; i > 0; i--) {
        int j = rand() % (i + 1);
        int tmp = arr[i];
        arr[i] = arr[j];
        arr[j] = tmp;
    }
}

/* 递归辅助：从 (row,col) 开始填充，候选数随机打乱 */
static bool fill_board_recursive(Cell board[BOARD_SIZE][BOARD_SIZE], int row, int col) {
    if (row == BOARD_SIZE) return true;                  /* 全部填完 */
    if (col == BOARD_SIZE) return fill_board_recursive(board, row + 1, 0);

    int nums[9] = {1, 2, 3, 4, 5, 6, 7, 8, 9};
    shuffle(nums, 9);

    for (int i = 0; i < 9; i++) {
        if (is_valid_placement(board, row, col, nums[i])) {
            board[row][col].value = nums[i];
            if (fill_board_recursive(board, row, col + 1)) return true;
            board[row][col].value = 0;                   /* 回溯 */
        }
    }
    return false;
}

/* 生成完整终盘（回溯 + 随机候选数顺序） */
void generate_full_board(Cell board[BOARD_SIZE][BOARD_SIZE]) {
    for (int r = 0; r < BOARD_SIZE; r++)
        for (int c = 0; c < BOARD_SIZE; c++)
            board[r][c].value = 0;

    fill_board_recursive(board, 0, 0);
}

/* 挖洞法生成题目 */
void generate_puzzle(GameState &state, int holes) {
    /* 1. 生成完整终盘作为解 */
    generate_full_board(state.solution);

    /* 2. 复制到玩家棋盘 */
    for (int r = 0; r < BOARD_SIZE; r++) {
        for (int c = 0; c < BOARD_SIZE; c++) {
            state.board[r][c].value     = state.solution[r][c].value;
            state.board[r][c].is_given  = true;
            state.board[r][c].is_conflict = false;
        }
    }

    /* 3. 随机挖洞 */
    /* 创建坐标数组并打乱 */
    int positions[81][2];
    for (int i = 0; i < 81; i++) {
        positions[i][0] = i / BOARD_SIZE;
        positions[i][1] = i % BOARD_SIZE;
    }
    /* Fisher-Yates 打乱坐标 */
    for (int i = 80; i > 0; i--) {
        int j = rand() % (i + 1);
        int tr = positions[i][0], tc = positions[i][1];
        positions[i][0] = positions[j][0];
        positions[i][1] = positions[j][1];
        positions[j][0] = tr;
        positions[j][1] = tc;
    }

    int dug = 0;
    for (int i = 0; i < 81 && dug < holes; i++) {
        int r = positions[i][0];
        int c = positions[i][1];
        int backup = state.board[r][c].value;

        state.board[r][c].value = 0;

        /* 验证仍有唯一解 */
        Cell temp[BOARD_SIZE][BOARD_SIZE];
        for (int rr = 0; rr < BOARD_SIZE; rr++)
            for (int cc = 0; cc < BOARD_SIZE; cc++)
                temp[rr][cc].value = state.board[rr][cc].value;

        if (count_solutions(temp, 2) == 1) {
            /* 唯一解，保留挖洞 */
            state.board[r][c].is_given = false;
            dug++;
        } else {
            /* 多解或无解，恢复 */
            state.board[r][c].value = backup;
        }
    }

    /* 更新空格计数 */
    state.empty_count = 0;
    for (int r = 0; r < BOARD_SIZE; r++)
        for (int c = 0; c < BOARD_SIZE; c++)
            if (state.board[r][c].value == 0)
                state.empty_count++;
}

/* ========== 游戏逻辑 ========== */

void init_game(GameState &state, int difficulty) {
    memset(&state, 0, sizeof(state));
    state.difficulty   = difficulty;
    state.cursor_row   = 4;
    state.cursor_col   = 4;
    state.game_over    = false;

    int holes = DIFFICULTIES[difficulty].holes;
    generate_puzzle(state, holes);

    /* solution 不需要 is_given 标记，仅 value 有效即可 */
    for (int r = 0; r < BOARD_SIZE; r++)
        for (int c = 0; c < BOARD_SIZE; c++)
            state.solution[r][c].is_given = false;
}

void place_number(GameState &state, int num) {
    int r = state.cursor_row;
    int c = state.cursor_col;

    /* 给定格不可修改 */
    if (state.board[r][c].is_given) return;

    /* 填入相同数字视为无操作 */
    if (state.board[r][c].value == num) return;

    /* 更新空格计数 */
    if (state.board[r][c].value == 0 && num != 0)
        state.empty_count--;
    else if (state.board[r][c].value != 0 && num == 0)
        state.empty_count++;

    state.board[r][c].value = num;

    /* 更新冲突标记 */
    update_conflicts(state);

    /* 检查胜利 */
    if (check_win(state)) {
        state.game_over = true;
    }
}

void erase_cell(GameState &state) {
    place_number(state, 0);
}

void update_conflicts(GameState &state) {
    for (int r = 0; r < BOARD_SIZE; r++) {
        for (int c = 0; c < BOARD_SIZE; c++) {
            if (state.board[r][c].value == 0) {
                state.board[r][c].is_conflict = false;
                continue;
            }
            state.board[r][c].is_conflict =
                !is_valid_placement(state.board, r, c, state.board[r][c].value);
        }
    }
}

bool check_win(const GameState &state) {
    /* 还有空格则未完成 */
    if (state.empty_count > 0) return false;

    /* 检查所有格子无冲突 */
    for (int r = 0; r < BOARD_SIZE; r++)
        for (int c = 0; c < BOARD_SIZE; c++)
            if (state.board[r][c].value == 0 || state.board[r][c].is_conflict)
                return false;

    return true;
}

/* ========== 渲染 ========== */

#define COLOR_RESET "\033[0m"

/* 获取格子的显示颜色码 */
static const char *cell_color(const GameState &state, int r, int c) {
    bool is_cursor = (r == state.cursor_row && c == state.cursor_col);

    if (state.board[r][c].is_conflict) {
        return is_cursor ? "\033[41;37m" : "\033[31m";
    }
    if (state.board[r][c].is_given) {
        return is_cursor ? "\033[44;1;37m" : "\033[1;37m";
    }
    if (state.board[r][c].value != 0) {
        return is_cursor ? "\033[44;1;36m" : "\033[36m";
    }
    return is_cursor ? "\033[44m" : "\033[0m";
}

void render_board(const GameState &state) {
    /* 列标题 */
    printf("    1 2 3   4 5 6   7 8 9\n");
    /* 顶部边框 */
    printf("  ┌───────┬───────┬───────┐\n");

    for (int r = 0; r < BOARD_SIZE; r++) {
        if (r > 0 && r % BOX_SIZE == 0)
            printf("  ├───────┼───────┼───────┤\n");

        printf("%d │", r + 1);
        for (int c = 0; c < BOARD_SIZE; c++) {
            if (c > 0 && c % BOX_SIZE == 0) printf(" │");

            const char *color = cell_color(state, r, c);
            if (state.board[r][c].value == 0)
                printf(" %s·" COLOR_RESET, color);
            else
                printf(" %s%d" COLOR_RESET, color, state.board[r][c].value);

            if (c < BOARD_SIZE - 1) printf(" ");
        }
        printf(" │\n");
    }

    printf("  └───────┴───────┴───────┘\n");
}

void render_status(const GameState &state) {
    printf("\n  剩余空格: %d\n", state.empty_count);
    printf("  方向键:移动  |  1-9:填入  |  0/Del:擦除  |  R:重新开始  |  Q:退出\n");
}

void render(const GameState &state) {
    printf("\033[H\033[J"); /* 回左上角并清至屏尾 */
    printf("  数独 SUDOKU - %s\n\n", DIFFICULTIES[state.difficulty].name);
    render_board(state);
    render_status(state);
    fflush(stdout);
}

File: engineering/apps/web/games-mini-program/src/utils/sudoku.ts
/**
 * @file sudoku.ts
 * @brief 数独游戏核心逻辑（纯 JS 实现）
 *
 * PC 终端版使用 C 代码：engineering/apps/games/sudoku/
 * 微信小程序版使用本 JS 实现
 */

export const SUDOKU_CONFIG = {
  GRID_SIZE: 9,
  BOX_SIZE: 3,
  DIFFICULTIES: [
    { name: '简单', holes: 30, maxTime: 180, description: '入门' },
    { name: '中等', holes: 40, maxTime: 300, description: '进阶' },
    { name: '困难', holes: 50, maxTime: 600, description: '挑战' }
  ]
}

// 单元格
export interface Cell {
  value: number      // 0=空, 1-9=数字
  isGiven: boolean   // 是否为题目固定格
  isConflict: boolean // 是否冲突
}

// 游戏状态
export interface SudokuState {
  board: Cell[][]           // 当前棋盘
  solution: number[][]      // 完整解
  cursorRow: number         // 光标行 (0-8)
  cursorCol: number         // 光标列 (0-8)
  difficulty: number        // 难度等级 (0-2)
  emptyCount: number        // 剩余空格数
  isGameOver: boolean       // 是否完成
  selectedNumber: number    // 当前选中的数字 (1-9)
  hintCount: number         // 剩余提示次数
}

/**
 * 创建空棋盘
 */
function createEmptyBoard(): Cell[][] {
  return Array(SUDOKU_CONFIG.GRID_SIZE)
    .fill(null)
    .map(() =>
      Array(SUDOKU_CONFIG.GRID_SIZE)
        .fill(null)
        .map(() => ({
          value: 0,
          isGiven: false,
          isConflict: false
        }))
    )
}

/**
 * 深拷贝棋盘
 */
function copyBoard(board: Cell[][]): Cell[][] {
  return board.map(row => row.map(cell => ({ ...cell })))
}

/**
 * 深拷贝解
 */
function copySolution(solution: number[][]): number[][] {
  return solution.map(row => [...row])
}

/**
 * 检查在某位置放置数字是否有效
 */
function isValidPlacement(
  board: Cell[][],
  row: number,
  col: number,
  num: number
): boolean {
  const { GRID_SIZE, BOX_SIZE } = SUDOKU_CONFIG

  // 检查行
  for (let c = 0; c < GRID_SIZE; c++) {
    if (c !== col && board[row][c].value === num) return false
  }

  // 检查列
  for (let r = 0; r < GRID_SIZE; r++) {
    if (r !== row && board[r][col].value === num) return false
  }

  // 检查 3x3 宫格
  const boxRow = Math.floor(row / BOX_SIZE) * BOX_SIZE
  const boxCol = Math.floor(col / BOX_SIZE) * BOX_SIZE
  for (let r = boxRow; r < boxRow + BOX_SIZE; r++) {
    for (let c = boxCol; c < boxCol + BOX_SIZE; c++) {
      if (r !== row && c !== col && board[r][c].value === num) return false
    }
  }

  return true
}

/**
 * 求解数独（回溯算法）
 */
function solveSudoku(board: Cell[][]): boolean {
  const { GRID_SIZE } = SUDOKU_CONFIG

  for (let row = 0; row < GRID_SIZE; row++) {
    for (let col = 0; col < GRID_SIZE; col++) {
      if (board[row][col].value === 0) {
        for (let num = 1; num <= GRID_SIZE; num++) {
          if (isValidPlacement(board, row, col, num)) {
            board[row][col].value = num
            if (solveSudoku(board)) return true
            board[row][col].value = 0
          }
        }
        return false
      }
    }
  }
  return true
}

/**
 * 统计解的数量（最多统计到 limit）
 */
function countSolutions(
  board: Cell[][],
  solution: number[][],
  limit: number = 2
): number {
  const { GRID_SIZE } = SUDOKU_CONFIG
  let count = 0

  // 将 board 复制到临时棋盘
  const tempBoard = copyBoard(board)

  for (let row = 0; row < GRID_SIZE; row++) {
    for (let col = 0; col < GRID_SIZE; col++) {
      if (tempBoard[row][col].value === 0) {
        for (let num = 1; num <= GRID_SIZE; num++) {
          if (isValidPlacement(tempBoard, row, col, num)) {
            tempBoard[row][col].value = num
            count += countSolutions(tempBoard, solution, limit - count)
            tempBoard[row][col].value = 0
            if (count >= limit) return count
          }
        }
        return count
      }
    }
  }

  // 找到一个解
  if (count === 0) {
    // 保存解
    for (let r = 0; r < GRID_SIZE; r++) {
      for (let c = 0; c < GRID_SIZE; c++) {
        solution[r][c] = tempBoard[r][c].value
      }
    }
    return 1
  }

  return count
}

/**
 * 生成完整数独解
 */
function generateFullBoard(): number[][] {
  const board = createEmptyBoard()
  solveSudoku(board)
  return board.map(row => row.map(cell => cell.value))
}

/**
 * 移除数字形成题目
 */
function removeNumbers(
  solution: number[][],
  board: Cell[][],
  holes: number
): number {
  const { GRID_SIZE } = SUDOKU_CONFIG
  const positions: { row: number; col: number }[] = []

  // 生成所有位置
  for (let r = 0; r < GRID_SIZE; r++) {
    for (let c = 0; c < GRID_SIZE; c++) {
      positions.push({ row: r, col: c })
    }
  }

  // 随机打乱
  for (let i = positions.length - 1; i > 0; i--) {
    const j = Math.floor(Math.random() * (i + 1))
    ;[positions[i], positions[j]] = [positions[j], positions[i]]
  }

  let removed = 0
  for (const pos of positions) {
    if (removed >= holes) break

    const { row, col } = pos
    const tempValue = solution[row][col]

    // 尝试移除
    board[row][col].value = 0

    // 检查是否还有唯一解
    const tempSolution: number[][] = Array(GRID_SIZE)
      .fill(null)
      .map(() => Array(GRID_SIZE).fill(0))

    if (countSolutions(board, tempSolution, 2) === 1) {
      removed++
    } else {
      // 恢复
      board[row][col].value = tempValue
    }
  }

  return removed
}

/**
 * 创建初始游戏状态
 */
export function createInitialState(difficulty: number = 0): SudokuState {
  const { GRID_SIZE } = SUDOKU_CONFIG
  const holes = SUDOKU_CONFIG.DIFFICULTIES[difficulty].holes

  // 生成完整解
  const solution = generateFullBoard()

  // 创建题目
  const board = createEmptyBoard()
  for (let r = 0; r < GRID_SIZE; r++) {
    for (let c = 0; c < GRID_SIZE; c++) {
      board[r][c].value = solution[r][c]
      board[r][c].isGiven = true
    }
  }

  // 挖洞
  const removed = removeNumbers(solution, board, holes)

  return {
    board,
    solution: copySolution(solution),
    cursorRow: 0,
    cursorCol: 0,
    difficulty,
    emptyCount: removed,
    isGameOver: false,
    selectedNumber: 1,
    hintCount: 3  // 初始 3 次提示
  }
}

/**
 * 放置数字
 */
export function placeNumber(state: SudokuState, num: number): void {
  const { cursorRow, cursorCol } = state
  const cell = state.board[cursorRow][cursorCol]

  // 固定格不能修改
  if (cell.isGiven) return

  // 放置数字
  cell.value = num

  // 更新冲突标记
  updateConflicts(state)

  // 检查是否完成
  checkWin(state)
}

/**
 * 擦除当前格子
 */
export function eraseCell(state: SudokuState): void {
  const { cursorRow, cursorCol } = state
  const cell = state.board[cursorRow][cursorCol]

  if (cell.isGiven) return

  cell.value = 0
  cell.isConflict = false

  // 重新计算空格子数
  recalcEmptyCount(state)
}

/**
 * 更新所有单元格的冲突标记
 */
export function updateConflicts(state: SudokuState): void {
  const { GRID_SIZE } = SUDOKU_CONFIG

  // 重置所有冲突
  for (let r = 0; r < GRID_SIZE; r++) {
    for (let c = 0; c < GRID_SIZE; c++) {
      state.board[r][c].isConflict = false
    }
  }

  // 检查每个已填格子
  for (let r = 0; r < GRID_SIZE; r++) {
    for (let c = 0; c < GRID_SIZE; c++) {
      const value = state.board[r][c].value
      if (value === 0) continue

      // 检查行
      for (let cc = 0; cc < GRID_SIZE; cc++) {
        if (cc !== c && state.board[r][cc].value === value) {
          state.board[r][c].isConflict = true
          state.board[r][cc].isConflict = true
        }
      }

      // 检查列
      for (let rr = 0; rr < GRID_SIZE; rr++) {
        if (rr !== r && state.board[rr][c].value === value) {
          state.board[r][c].isConflict = true
          state.board[rr][c].isConflict = true
        }
      }
    }
  }
}

/**
 * 重新计算空格子数
 */
function recalcEmptyCount(state: SudokuState): void {
  const { GRID_SIZE } = SUDOKU_CONFIG
  let count = 0
  for (let r = 0; r < GRID_SIZE; r++) {
    for (let c = 0; c < GRID_SIZE; c++) {
      if (state.board[r][c].value === 0) count++
    }
  }
  state.emptyCount = count
}

/**
 * 检查是否胜利
 */
export function checkWin(state: SudokuState): void {
  const { GRID_SIZE } = SUDOKU_CONFIG

  for (let r = 0; r < GRID_SIZE; r++) {
    for (let c = 0; c < GRID_SIZE; c++) {
      if (state.board[r][c].value === 0) return
    }
  }

  state.isGameOver = true
}

/**
 * 移动光标
 */
export function moveCursor(state: SudokuState, dx: number, dy: number): void {
  const { GRID_SIZE } = SUDOKU_CONFIG
  state.cursorCol = (state.cursorCol + dx + GRID_SIZE) % GRID_SIZE
  state.cursorRow = (state.cursorRow + dy + GRID_SIZE) % GRID_SIZE
}

/**
 * 设置光标位置
 */
export function setCursor(state: SudokuState, row: number, col: number): void {
  state.cursorRow = Math.max(0, Math.min(8, row))
  state.cursorCol = Math.max(0, Math.min(8, col))
}

/**
 * 选中数字
 */
export function selectNumber(state: SudokuState, num: number): void {
  state.selectedNumber = Math.max(1, Math.min(9, num))
}

/**
 * 提示（显示一个正确数字）
 */
export function hint(state: SudokuState): boolean {
  const { GRID_SIZE } = SUDOKU_CONFIG

  // 检查提示次数
  if (state.hintCount <= 0) {
    return false  // 提示次数已用完
  }

  // 找到第一个空格
  for (let r = 0; r < GRID_SIZE; r++) {
    for (let c = 0; c < GRID_SIZE; c++) {
      if (state.board[r][c].value === 0) {
        state.board[r][c].value = state.solution[r][c]
        state.board[r][c].isGiven = true
        state.emptyCount--
        state.hintCount--  // 消耗一次提示
        updateConflicts(state)
        checkWin(state)
        return true
      }
    }
  }
  return false
}